docs/enhancements/consumers-scheduling.md

```diff
@@ -0,0 +1,111 @@
+	MaxAllowedVReplicas *int32      `json:"maxAllowedVReplicas,omitempty"`
+	Placement           []Placement `json:"placements,omitempty"`
+}
+
+type Placement struct {
+	// PodName is the name of the pod where the resource is placed
+	PodName string `json:"podName,omitempty"`
+	// VReplicas is the number of virtual replicas assigned to in the pod
+	VReplicas int32 `json:"vreplicas,omitempty"`
+}
+```
+
+The `KafkaSource` type embeds the `Placeable` type in the status, while the Trigger will use
+a different representation of that type in the `status.annotations` field like the following:
+
+```yaml
+apiVersion: eventing.knative.dev/v1
+kind: Trigger
+# ...
+metadata:
+  annotations:
+    # This is set by the Trigger author to signal how many consumers they want to run.
+    kafka.eventing.knative.dev/replicas: <num_replicas>
+status:
+  annotations:
+    # Embed placeable type in status annotations: 
+    # - https://github.com/knative-sandbox/eventing-kafka/blob/main/pkg/apis/duck/v1alpha1/placement_types.go
+    kafka.eventing.knative.dev/scheduler.placeable.MaxAllowedVReplicas: num_topic_partitions
+    kafka.eventing.knative.dev/scheduler.placeable.Placements.0.PodName: kafka-broker-dispatcher-abc
+    kafka.eventing.knative.dev/scheduler.placeable.Placements.0.VReplicas: <num_replicas_for_pod_kafka-broker-dispatcher-abc>
+    kafka.eventing.knative.dev/scheduler.placeable.Placements.1.PodName: kafka-broker-dispatcher-xyz
+    kafka.eventing.knative.dev/scheduler.placeable.Placements.1.VReplicas: <num_replicas_for_pod_kafka-broker-dispatcher-xyz>
+    # ...
+```
+
+Once a resource reconciler made a scheduling decision a new reconciler will actually schedule resources to dispatcher pods.
+We call this new reconciler the [`Pod` reconciler](#Pod-Reconciler) since it will reconcile `corev1.Pod`s.
+
+## Resource Reconcilers
+
+Resource reconcilers are the `Trigger`, `KafkaSource`, and `KafkaChannel` reconciler.
+
+High level flow of the resource reconcilers will be:
+
+1. Schedule the consumer replicas into available pods by updating the `Placeable` status
+1. Reconcile the resource `Spec` (addressable sink, dead letter sink, etc)
+1. Mark the resource ready when the `Pod` reconciler has seen the updated scheduling decision
+
+## Pod Reconciler
+
+The pod reconciler will reconcile `dispatcher` pods.
+
+A pod will be reconciled when a `placeable`, that references a pod, changes.
+
+The high level flow of the `Pod` recoconciler is:
+
+1. Get all resources (Trigger or KafkaSource) assigned to it
+1. Transform them to the internal contract representation
+1. Save the contract in the ConfigMap associtated with the pod
+1. Update pod annotations as follow:
+    ```yaml
+    apiVersion: v1
+    kind: Pod
+    metadata:
+    annotations:
+        # <resource_namespace>/<resource_name> represents the resource key
+        kafka.eventing.knative.dev/scheduler.placeable.Placements.<resource_namespace>/<resource_name>: <resource.status.ObservedGeneration>
```

Comment:
How will this look like if there are a significant number of Triggers, specifically:
1. The UX: Are we okay for cluster admins dealing with a pod that has 100+ annotations? 
2. Are we gonna hit any [etcd limits](https://etcd.io/docs/v3.5/dev-guide/limit/#request-size-limit)?

Reply:
> How will this look like if there are a significant number of Triggers, specifically:
> 
> 1. The UX: Are we okay for cluster admins dealing with a pod that has 100+ annotations?

In general, I don't think that would be a problem, given that an annotation is not indexed while labels are, so it's comparable to additional fields of the spec. 

Besides that, I'm thinking that the best thing you want to do in Kubernetes is to have "low" resource limits and requests (whatever that means) for each pod and scale them out (if the application allows that) in order to take advantage to the high-density scheduling of the k8s pods, which leads to a lower number of annotations per pod.

> 2. Are we gonna hit any [etcd limits](https://etcd.io/docs/v3.5/dev-guide/limit/#request-size-limit)?

In the worst-case, in addition to other pod's fields, the `kafka.eventing.knative.dev/scheduler.placeable.Placements.<resource_namespace>/<resource_name>` (when variables are 63 chars) is 740 bytes x 100 = 74000 bytes, which shouldn't be a problem for Etcd or the API server.

Reply:
> > > > In general, I don't think that would be a problem, given that an annotation is not indexed while labels are, so it's comparable to additional fields of the spec.
> > > 
> > > 
> > > Can you elaborate more why is annotation indexing relevant here?
> > 
> > 
> > It's not that relevant because it's not done. (forget about what I wrote :) )
> > > > Besides that, I'm thinking that the best thing you want to do in Kubernetes is to have "low" resource limits and requests (whatever that means) for each pod and scale them out (if the application allows that) in order to take advantage to the high-density scheduling of the k8s pods, which leads to a lower number of annotations per pod.
> > > 
> > > 
> > > This is still assuming a specific user behavior, the question is, given that we're using this in a serverless environment where scale to 0 of subscribers is assumed, is the use case of say two dispatcher pods handling 1K triggers totally out of question here? and how would the admin UX be like in that case?
> > 
> > 
> > If you scale all triggers to 0, we will have 0 scheduling-relevant annotations on both pods.
> 
> Is there a misunderstanding here? I'm referring to scaling of the _subscriber_ of the trigger, not the trigger itself.

Sorry for the misunderstanding, but then my comment below is the answer.

> 
> > We have a new annotation for each trigger that has `replicas > 0`.
> > Assuming that all 1k triggers have `replicas > 0` and we have only 2 pods, then yes 500 annotations per pod.
> > (2 pods for 1k Kafka consumers is definitely low)
> 
> I'm interested honestly to know more concrete numbers or sizing guides behind that, care to elaborate more?
> 

I don't have concrete numbers but scheduling 500 consumers in a pod means that if one of them is rescheduled you will get a rebalance across multiple consumer groups (when there are more consumers for a group in the same pod the rebalance could even be cascading (one join the group, then another one join group, and so on) which leads to a lot of spike traffic to Kafka and message processing delays since joining a group is a stateful operation that needs to be coordinated.

Don't you see an underprovisioned environment in running 500 Kafka consumers in a single process?

> > I think that since those are admin decisions, an admin needs to know how the system works to best operate it, so I'd say spelling out in docs that the preference is for small pods over big pods is a reasonable solution IMHO.
> 
> I can totally be convinced with that when it's unavoidable, but is having a dynamically growing list of annotations in the pod unavoidable? I think that's what we should focus on. If it is, sure let's fall back to this solution.

A max trigger per pod configuration could be introduced but that is not relevant for scheduling that is an autoscaling problem because once we have reached the capacity of a pod we need to add more pods (or remove) and then the scheduler will do its own things as described here.

Reply:
> > I don't have concrete numbers but scheduling 500 consumers in a pod means that if one of them is rescheduled you will get a rebalance across multiple consumer groups (when there are more consumers for a group in the same pod the rebalance could even be cascading (one join the group, then another one join group, and so on)
> 
> I need to think more of this, quick question, are vertx consumers "verticles" ?
> 

Yes, every KafkaConsumer instance is wrapped in a vertx verticle

> > A max trigger per pod configuration could be introduced but that is not relevant for scheduling that is an autoscaling problem because once we have reached the capacity of a pod we need to add more pods (or remove) and then the scheduler will do its own things as described here.
> 
> Why do we need an annotation entry per trigger?

To coordinate with the `Trigger` reconciler (queue the trigger, verifying the observed generation, etc)

Reply:
> To coordinate with the `Trigger` reconciler (queue the trigger, verifying the observed generation, etc)

I'm sorry I meant on the pod resource, why an annotation entry per trigger?